### ISSUE TITLE
"Undo" in Fluid is not able to use

### ISSUE THREAD

**Issue by traditionalgithub** [fltk/fltk#84]
"Undo" in Fluid is not able to use, and is consistently grayed. Built on Ubuntu (an Ubuntu equivalent to be specific) 20.04 with gcc-8.

**Comment by Albrecht-S** [comment]
Thanks for the report. I can confirm that this is a regression introduced in d6294ac2fccab7e7ffdfbf1ef63a059bba46c777 - at least as far as `git bisect` reports:
```
$ git bisect good
d6294ac2fccab7e7ffdfbf1ef63a059bba46c777 is the first bad commit
commit d6294ac2fccab7e7ffdfbf1ef63a059bba46c777
Author: Matthias M ... (edited)
Date:   Fri Feb 1 23:16:49 2019 +0100

    Reorganized Fluid Template feature (STR #3336).
```

Test scenario: open a fluid file (e.g. `test/valuators.fl`), delete one or more widgets, then try to use `undo` (i.e. `ctrl-z`) for a quick test. I can also confirm that the `undo` menu entry is inactive (grayed out).

Unfortunately the changes introducing this bug are not obvious to me (right now). I need to take a deeper look into it. There are many more - maybe related - commits to fluid around this one, so the found commit may also be a false positive (not very likely though).

I'm going to work on this soon unless someone else (@MatthiasWM maybe?) has a quick solution.

**Comment by erco77** [comment]
As I wrote on fltk.general (I wasn't aware a bug had been opened):
---
Confirmed.
This certainly looks suspicious:

`undo.cxx:#define UNDO_ITEM	25		// Undo menu item index`

I'm not even sure if that's the problem, it just looks wrong to have a fixed index into the main menu; the fluid menu can change in size (File -> [[recent files]]) and there's probably been additions to the
fluid menu due to feature growth. [If that's the case, probably no one realized changing the menu size would break anything.]

I'd offer that fixed index should be replaced with something more dynamic-proof, e.g. calling

		`undo_index = Fl_Menu_::find_index(undo_cb);`

[should probably check for similar problems elsewhere in fluid]

**Comment by erco77** [comment]
Suggesting this patch -- I gave it a quick check, seems to restore undo/redo functionality.

[fluid.patch.zip](https://github.com/fltk/fltk/files/4673751/fluid.patch.zip)

Sorry about the zip; seems github issues won't let you load a .patch file : /

**Comment by erco77** [comment]
Pretty confident about the patch being necessary; committed dcfe826

**Comment by erco77** [comment]
OP confirmed fix, closing. Albrecht, if you have further comments, feel free to re-open.

**Comment by Albrecht-S** [comment]
@erco77 Great, you found the culprit. I was just testing after a remote help session with someone else and I definitely found that the `undo` and `redo` menu item(s) were off by one. I agree that your solution to find the menu item is much better than using constants.

Reopening, just to be sure: Greg, did you see [Manolo's comment](https://github.com/fltk/fltk/commit/dcfe826b2f06144cf54ce4f1a93bccf1b40381d9#commitcomment-39401926) to your commit? I believe he's correct in that this was maybe not intentional (unless you fixed another bug).

I'm going to take a closer look later, maybe you can confirm that it was intentional or fix the issue?

**Comment by Albrecht-S** [comment]
(1) @erco77: I'm pretty sure that @ManoloFLTK's observation is correct and that the new code is wrong in one place. After `redo`ing everything the redo menu should be deactivated. I can commit the fix if requested, tested and works AFAICT.

(2) Another loosely related observation is that both `undo` and `redo` menus are active at program start. This is no regression (it's been so even in 1.3) but it could be "fixed" as well, if only to be consistent. I believe that both menus should be deactivated at program start and/or after loading a new file.

I'm awaiting a comment on (1) and investigating (2) right now...

**Comment by Albrecht-S** [comment]
Fixed (both 1 and 2) in commit adb2a719d4.

Please test and confirm. I believe this issue can be closed now.

**Comment by erco77** [comment]
Will check the new additions.

Glad you caught my apparent search/replace screwup in redo_cb() - sigh.

And regarding undo_clear() -- while I did notice that when you run fluid, undo/redo items weren't grayed out like they should be, but I didn't try to fix that, as I thought it a separate issue. But glad you guys fixed that by tweaking undo_clear(). I believe that problem has always been in fluid.

**Comment by erco77** [comment]
Looks good to me.
The only thing semi-related that I think might be odd is you can undo a "New", but can't undo a File -> Open.
I'm not really sure which is 'right'; I can see wanting to undo an accidental 'New' (the "Do you want to save first" dialog should prevent one from making that mistake, unless you accidentally type past the dialog), but I can also see doing some working, and only expecting the undo buffer to be for the current project, and not resurrect old projects.

**Comment by Albrecht-S** [comment]
Yep, I noticed the New / Redo and/or File Open / Redo behavior "somehow" while testing, but I didn't (want to) dig deeper into this. The menu (de)activation issue seemed directly related and since I could spend some time I did it. It's probably not perfect (too much code duplication), but it seems to work.

I'd tend to close this issue and leave the open/undo/redo stuff as-is for now if you don't mind. I'm also not sure what the 'right' behavior would (should) be.
